### PR TITLE
feat: Add support for cloning Google Cloud Source repos (#7534)

### LINKED ISSUE #7534: Add support for Google Cloud Source repositories
# Summary

This enhancement aims at adding support for Google Cloud Source repositories. There are already a couple of issues (see #3379 and #2814), where the overall consensus was that it is currently not possible or experimental at best. Also, there is no documented and easy to use solution to clone Google Cloud Source repositories.

# Motivation

ArgoCD users with repositories hosted on Google Cloud Source are currently unable to use these repositories with ArgoCD. Adding support for such repositories would allow such users to use ArgoCD as well.

# Proposal

I am proposing support for Google Cloud Service Account credentials in JSON format as part of one of the ArgoCD repository or repository credential secrets, which can be used to create access tokens to clone Google Cloud Source repositories.

I already have a locally tested solution for this. I am planning to refactor a couple of bits, create proper documentation, argocd-cli support, etc. and then to create a PR referencing this enhancement proposal.

### PR DESCRIPTION
**Pull request by david-becher**
This PR closes #7534. I adjusted both the cli as well as the UI to add support for using private Google Cloud Source repos.

I welcome any feedback on this PR, since I am a first-time contributor.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by tantweiler** [comment]
@alexmt did you have the chance for a review? We would really like to have this feature in master soon. 

Thank you in advance!

**Comment by PavelKarpenka** [comment]
This is exactly what we were looking for! 
@david-becher do you know, is there any chance to see this in master?

**Comment by david-becher** [comment]
Hi @PavelKarpenka,

I would love to see this in master as well, unfortunately no one has gotten around yet to review my PR yet. :(

And as more and more time passes, I always need to rebuild it and resolve merge conflicts in protbuf-related files and then ask for GitHub action approval, so that tests can be run.

@alexmt Is there any way you can signal me, when you or anyone else will be having time for review? Then I would update my PR so it does not have never conflicts with protobuf files anymore. But I don't see the point in updating the PR when it's not clear when this is getting reviewed anyway...

**Comment by cedricOL** [comment]
@alexmt any chance to review this feature?

**Comment by tantweiler** [comment]
Sorry guys but this behavior is a bit annoying for people who would like to start contributing to open source projects. If someone is busy...fine. But not giving an answer for months is a no go in my opinion. Isn't there another maintainer available where your @alexmt can forward the review request? There are many people who are interested in that PR which was provided by David and we would like to see that in master pretty soon if the new checks will all pass. Thank you!

**Comment by alexmt** [comment]
Sorry about that experience @tantweiler :( We are overloaded. Working on adding more reviewers and approvers but it takes time. I will review PR ASAP

**Review by jannfis — CHANGES_REQUESTED** [review]
Thanks a lot for this PR @david-becher. and apologies for forgetting about it so long. With the current amount of PRs and issues coming in, it's not easy to keep an overview over things that are not on our (the maintainers) priority lists and that take significant effort to test (e.g. access to Google cloud services).

I have not tested the changes in lack of a proper testing environment, but the code looks good to me so far with some concerns addressed below. PTAL.

And just as a heads-up: We do have regular (weekly) contributor meetings where things such as pending PRs can be addressed directly. The next time one of your PRs is pending review for a while, feel free to chime in and bring it to our direct attention :) Details can be found in the [Argoproj Community Calendar](https://calendar.google.com/calendar/embed?src=argoproj@gmail.com)

**Comment by david-becher** [comment]
Hi @jannfis

Thanks for getting back to me. I will be having time in the following week to address the comments you made to the code changes. :)

> And just as a heads-up: We do have regular (weekly) contributor meetings where things such as pending PRs can be addressed directly. The next time one of your PRs is pending review for a while, feel free to chime in and bring it to our direct attention :) Details can be found in the [Argoproj Community Calendar](https://calendar.google.com/calendar/embed?src=argoproj@gmail.com)

I did not know that, but I will make a mental note of it in case for future code contributions. :)

File: util/git/creds.go [hunk 1]
```diff
@@ -378,3 +386,85 @@ func (g GitHubAppCreds) GetClientCertData() string {
...
+		ClientEmail  string `json:"client_email"`
+		PrivateKeyID string `json:"private_key_id"`
+		PrivateKey   string `json:"private_key"`
+		AuthURL      string `json:"auth_uri"`
+		TokenURL     string `json:"token_uri"`
+		ProjectID    string `json:"project_id"`
+	}
+
+	var f googleCredentialsFile
+	if err := json.Unmarshal(c.creds.JSON, &f); err != nil {
```
[the remark this answers is not in the document]
Comment (david-becher) [reply util/git/creds.go:438]: Addressed with last push.

File: util/git/creds.go [hunk 2]
```diff
@@ -378,3 +386,85 @@ func (g GitHubAppCreds) GetClientCertData() string {
...
+		return "", err
+	}
+	return f.ClientEmail, nil
+}
+
+func (c GoogleCloudCreds) getAccessToken() (string, error) {
+	// Compute hash of creds for lookup in cache
+	h := sha256.New()
+	_, err := h.Write(c.creds.JSON)
```
[the remark this answers is not in the document]
Comment (david-becher) [reply util/git/creds.go:451]: Addressed with last push.

**Comment by g0dil** [comment]
Hi @david-becher, I really hope this can get reviewed soon since we would also like to make use of this feature.

If I read the code correctly, this change does not support application default credentials? This could easily be implemented in `NewGoogleCloudCreds` by adding `FindDefaultCredentials` as an alternative for `CredentialsFromJSON`. Several Ideas come to mind:
* try to parse the json as a simple object which just contains a single key, e.g. `useDefaultCredentials: true`. I would find this to be the preferred solution.
* when `CredentialsFromJSON` fails, try `FindDefaultCredentials` instead (probably not a good idea since we don´t get any error in this case).
* if json parsing is considered to complicated, just compare the json string against some magic constant string value. However I find the first suggestion much cleaner.

I would greatly appreciate any feedback. If there are other possibilities to use workload identity to authenticate against a google cloud repository, please let me know (I know, I can use workload identity federation but that's a different beast).

**Comment by mrmatimba** [comment]
Hi @alexmt @jannfis , is there any progress in reviewing this PR? I am eagerly waiting for it :)

**Comment by M0dj0** [comment]
I can only join, as I am urgently waiting for it.

**Comment by t1murl** [comment]
Hi @alexmt @jannfis , hope you guys are doing well.
Quite some time has passed since this PR has been created. Has someone found time to make progress with this? If this PR is missing something, I'm sure the community would appreciate a status comment so we can contribute if needed.
If availability is (still) the problem: I also would appreciate if this PR could get a little bit more love :).
Would it help to join and try to raise this in one of the project / contributor meetings that have been suggested above?
Thanks for any update on this.

**Comment by crenshaw-dev** [comment]
Apologies, this slipped through the cracks. If OP could resolve conflicts, or if someone can create a new PR based on this one and resolve conflicts, I am happy to review!

**Comment by alexef** [comment]
On it.

**Comment by crenshaw-dev** [comment]
@alexef someone took a crack at it, but I think they had to drop UI support due to gnarly conflicts and not being super up on React. https://github.com/argoproj/argo-cd/pull/11543

Maybe base on theirs and add UI support?

**Comment by alexef** [comment]
Closing in favor of #11618 (which merges this one with master). testers are welcome. 
So far I was able to use the UI to create the repository and check that it was correctly created as a secret.

**Pull request closed without merging**